internal/unionstore/pipelined_memdb.go

```
@@ -0,0 +1,159 @@
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing        atomic.Bool
+	errCh             chan error
+	flushFunc         FlushFunc
+	memBuffer         *MemDB
+	flushingMemBuffer *MemDB
+	len, size         int // len and size records the total flushed and onflushing memdb.
+}
+
+const (
+	MinFlushKeys = 10000
+	MinFlushSize = 16 * 1024 * 1024  // 16MB
+	MaxFlushSize = 128 * 1024 * 1024 // 128MB
+)
+
+type FlushFunc func(*MemDB) error
+
+func NewPipelinedMemDB(flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memBuffer: newMemDB(),
+		errCh:     make(chan error, 1),
+		flushFunc: flushFunc,
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memBuffer.Dirty() || p.flushingMemBuffer != nil
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemBuffer MemBuffer > Snapshot.
+func (p *PipelinedMemDB) Get(k []byte) ([]byte, error) {
+	v, err := p.memBuffer.Get(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		v, err = p.flushingMemBuffer.Get(k)
+	}
+	if err != nil {
+		return v, err
+	}
+	if len(v) == 0 {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
+
+func (p *PipelinedMemDB) GetFlags(k []byte) (kv.KeyFlags, error) {
+	f, err := p.memBuffer.GetFlags(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		f, err = p.flushingMemBuffer.GetFlags(k)
+	}
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+// MayFlush is called during execution of a transaction, it does flush when there are enough keys and the ongoing flushingMemBuffer is done.
+func (p *PipelinedMemDB) MayFlush() error {
+	if p.flushFunc == nil {
+		return nil
```

Shall we return an error here as no flushFunc is unexpected?